### ISSUE TITLE
does sageattention support bias?

### ISSUE THREAD

**Issue by pengzhangzhi** [thu-ml/SageAttention#111]
Hi,

I wonder if sageattention supports bias.  Like in `scaled_dot_product_attention(query, key, value, attn_mask=None, ...` 

attn_mask (optional Tensor) – Attention mask; shape must be broadcastable to the shape of attention weights, which is (N,...,L,S). Two types of masks are supported. A boolean mask where a value of True indicates that the element should take part in attention. A float mask of the same type as query, key, value that is added to the attention score.

**Comment by pengzhangzhi** [comment]
@rubbybbs @miskcoo @meta-inf @P2333  Would love to hear ur perspective and submit a pr if it's possible!

**Comment by miskcoo** [comment]
You should quote the main contributors of this repo. @jt-zhang @jason-huang03 .

**Comment by pengzhangzhi** [comment]
sorry lol i wasn't really knowing who is who. thx!

**Comment by jason-huang03** [comment]
Hi, adding attention bias would greatly reduce the speed. In this case, I think you can just use sdpa with attention bias.

**Comment by pengzhangzhi** [comment]
thanks! @jason-huang03 i've been trying to add the bias term. I wonder why adding the bias term will slow down? Say we have a bias of [B, H, N, N], we can just load the bias in the flash attn fwd loop, seems straightforward?

**Comment by zhiwei-dong** [comment]
Curious too.

**Comment by jason-huang03** [comment]
First, loading the bias consumes memory bandwidth. Second (this is important), it takes up registers. Actually attention kernel is register bound, so loading bias will cause register spill. Third, adding bias to the score takes up cuda core.

I believe the second is the most important aspect that will slow down the kernel.

**Comment by zhiwei-dong** [comment]
Get.

**Comment by pengzhangzhi** [comment]
> First, loading the bias consumes memory bandwidth. Second (this is important), it takes up registers. Actually attention kernel is register bound, so loading bias will cause register spill. Third, adding bias to the score takes up cuda core.
> 
> I believe the second is the most important aspect that will slow down the kernel.

thanks! I guess in the inference, adding bias will make it no difference than sdpa. 
But if we work on training attention wit bias, should it be faster given the fact that flash attn has a significant improvement in the training

**Comment by jason-huang03** [comment]
@pengzhangzhi I believe that if we need bias, we should separate the attention into sub-operations that use only causal/non-causal masks, or use the trick like right padding in language pretraining. This gives the highest throughput benefit.

In which case do we need to use attention bias in training?

**Comment by pengzhangzhi** [comment]
thanks! @jason-huang03 to ur question, I work on general ML not just llm where for example AlphaFold3 uses the attention with differentiable bias.
I implemented the flash attention with bias in triton for those who are interested!
https://github.com/pengzhangzhi/Flash-Attention-with-Bias-Triton
